agents/magpie/magpie_agent.py

```diff
@@ -0,0 +1,388 @@
+        import quaternionarray as qa
+        import toml
+
+        dets = toml.load(wafer_file)
+        fp = cls(ndets=len(dets))
+        band_idxs = {}
+        bands_seen = 0
+
+        xaxis = np.array([1., 0., 0.])
+        zaxis = np.array([0., 0., 1.])
+        cmaps = ["red_cmap", "blue_cmap"]
+
+        def det_coords(det):
+            quat =np.array(det['quat']).astype(np.float64)
+            rdir = qa.rotate(quat, zaxis).flatten()
+            ang = np.arctan2(rdir[1], rdir[0])
+            orient = qa.rotate(quat, xaxis).flatten()
+            polang = np.arctan2(orient[1], orient[0])
+            mag = np.arccos(rdir[2]) * 180 / np.pi
+            xpos = mag * np.cos(ang)
+            ypos = mag * np.sin(ang)
+            return (xpos, ypos), polang
+
+        for key, det in dets.items():
+            chan = det['channel']
+            (x, y), polangle = det_coords(det)
+
+            if det['band'] not in band_idxs:
+                band_idxs[det['band']] = bands_seen
+                bands_seen += 1
+            color_idx = band_idxs[det['band']]
+
+            fp.xs[chan] = x * wafer_scale
+            fp.ys[chan] = y * wafer_scale
+            fp.rots[chan] = polangle
+            fp.cnames[chan] = key
+            fp.eqs[chan] = f"/ + 1 s {key} 2"
+            fp.cmaps[chan] = cmaps[color_idx]
+            fp.templates[chan] = f"template_c{color_idx}_p0"
+
+        return fp
+
+
+
+class MagpieAgent:
+    mask_register = 'AMCc.SmurfProcessor.ChannelMapper.Mask'
+
+    def __init__(self, agent, args):
+        self.agent = agent
+        self._running = False
+
+        self.target_rate = args.target_rate
+        layout = args.layout.lower()
+        if layout == 'grid':
+            self.fp = FocalplaneConfig.grid(
+                args.xdim, args.ydim, ygap=8
+            )
+        elif layout == 'wafer':
+            if args.csv_file is not None:
+                self.fp = FocalplaneConfig.from_tune_csv(
+                    args.csv_file, wafer_scale=args.wafer_scale
+                )
+            else:
+                self.fp = FocalplaneConfig.from_wafer_file(
+                    args.wafer_file, wafer_scale=args.wafer_scale
+                )
+
+        self.sent_cfg = False
+        self.num_dets = self.fp.num_dets
+        self.mask = np.arange(self.num_dets)
+        self.mask = np.arange(4960)
+
+        self.reader = core.G3Reader(args.src)
+        self.sender = core.G3NetworkSender(
+            hostname='*', port=args.dest, max_queue_size=1000
+        )
+
+        self.lowpass = FIRFilter.butter_lowpass(self.target_rate, 200., order=4)
+        self.send_config_flag = True
+        self.scale = 1
+        self.offsets = np.zeros(self.fp.num_dets)
+        self.zero_flag = True
+
+        self.demodulate=True
+        self.demod_filter = FIRFilter.butter_lowpass(1, 200., order=4)
+        self.demod_freq = 4
+
+    def set_scale(self, session, params={}):
+        self.scale = params['scale']
```

Comment:
Personally I'd rather the task fail loudly when a required param is missing instead of choosing a default, so I normally just try to access the key and it'll raise KeyError if there's a problem. Do you think I should be handling this differently?

Reply:
Yeah, I still think it'd be good to handle the `KeyError`. If the proper param isn't provided you could then `return False, 'scale' is a required parameter` or similar to mark the task as failed.

Reply:
Would it be helpful if Agent code could raise some special OCS errors to propagate such failures outwards?  That would go along with some functions to help unpack parameters.  E.g. what if you could do:
```
    self.scale = get_required(params, 'scale')
```
and that would raise ocs.ParamMissingError('scale'), which would be caught by the OCSAgent wrapper and returned to the user as (False, "Parameter 'scale' is required.")

I think this approach would be in line with encouraging concise Agent code without too much boilerplate that still can propagate errors adequately.

Reply:
That sounds helpful to me. I've been pointing out the need to handle missing params in recent PRs, typically encouraging what I have here, using `.get()` with a sensible default value. It would definitely help in cases like this where one wants to force a parameter being set.